packages/grid/data-grid/src/tests/pagination.DataGrid.test.tsx

```diff
@@ -90,6 +91,34 @@ describe('<DataGrid /> - Pagination', () => {
       expect(onPageChange.callCount).to.equal(2);
     });
 
+    it('should paginate on trigger of onPageChange when clicking on next/previous page even when page prop is supplied', () => {
+      const TestDataGrid = () => {
+        const [page, setPage] = React.useState(1);
+
+        const handleOnPageChange = (params: GridPageChangeParams) => {
+          setPage(params.page);
+        };
+
+        return (
+          <div style={{ width: 300, height: 300 }}>
+            <DataGrid
+              {...baselineProps}
+              page={page}
+              onPageChange={handleOnPageChange}
```

Comment:
Instead of testing if the pagination works, could you use a `spy` here and check if `params.page` is the right value?

Reply:
The issue is not if `params.page` value is right or wrong. It's right even now in current release. The issue is that page prop seems to take precedence over `params.page` during changing of page.

Reply:
Actually, the issue is that the value of `params.page` is wrong. The idea of a `spy` is to check that the callback is called with the right params and that we'll not break this again in the future. If, instead of calling `setPage(params.page)`, I call `setPage(page > 1 ? page - 1 : page + 1)` the test will also be green but the params could be wrong.

https://codesandbox.io/s/material-demo-forked-w7ynh?file=/demo.js

![YEAl9QrVt6](https://user-images.githubusercontent.com/42154031/119141776-0de69d00-ba1c-11eb-9ae5-7e12a7465571.gif)

Reply:
@m4theushw Correct, I was about to make the same feedback. The current test is not as good as it could be. This #1711 issue is interesting as it surfaces how the previous logic was not following the convention. If the state is controlled, the internal state should never be the source of truth. I'm not sure how this bug should be fixed. In `GridPagination` we call `apiRef!.current!.setPage` with the new value. Maybe setPage should fire a `GRID_PAGESIZE_CHANGE` event (not `GRID_PAGESIZE_CHANGED`).

Reply:
> If the state is controlled, the internal state should never be the source of truth. I'm not sure how this bug should be fixed.

@oliviertassinari I didn't understand this statement. The `onPageChange` should be called with the new page number, so it can be used to update the `page` prop. It's the same as in a Select component, where the callback is always called with the new value, not the prop value.

> Maybe setPage should fire a GRID_PAGESIZE_CHANGE event (not GRID_PAGESIZE_CHANGED).

Yeah, this event is not following our convention. #1383

Reply:
> I didn't understand this statement.

I meant that I have seen controlled states that aren't. You could see the grid change the internal value, and the screen just after is reverted back by the parent of the grid.

Reply:
> Actually, the issue is that the value of params.page is wrong. The idea of a spy is to check that the callback is called with the right params and that we'll not break this again in the future. If, instead of calling setPage(params.page), I call setPage(page > 1 ? page - 1 : page + 1) the test will also be green but the params could be wrong.

@m4theushw You are correct. I was checking the internal `setPage` params at:
```js
  const setPage = React.useCallback(
    (page: number) => {
      logger.debug(`Setting page to ${page}`);
      console.log('Setting page to: ', page);
      setGridState((state) => ({
        ...state,
        pagination: applyConstraints(
          {
            ...state.pagination,
            page,
          },
          options.page,
        ),
      }));
      forceUpdate();
```

https://user-images.githubusercontent.com/20900032/119152554-09e36b00-ba6e-11eb-8018-9925f6ed796c.mov

But the event `GRID_PAGE_CHANGED` was publishing wrong `params` which gets published after the internal state through `applyConstraints` is updated. I will update the test. Thanks.

Reply:
Updated the test case.